### ISSUE TITLE
False positive for dereferencing null pointer (C++ 'this' pointer)

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#10818]
|  |  |
| --- | --- |
| Bugzilla Link | [10446](https://llvm.org/bz10446) |
| Version | unspecified |
| OS | MacOS X |
| Attachments | [test case](https://user-images.githubusercontent.com/60944935/143743514-7abf2468-f461-4743-af24-6d2260c8e7c6.gz) |
| Reporter | LLVM Bugzilla Contributor |

## Extended Description 

Using downloaded build checker-257, if I assign a variable the contents of 'this', then change that variable as if it were stepping along a linked list, I get a false positive dereference of a null pointer of 'this'.

The attached test case shows the original code (ifdefed out), plus a simplified version. Here is a snippet of it, but the attached file is a complete compilable demonstration. 'this' is never 0.
```c++
{
    const A *a = this;
    if (a != 0)
        a = a->m_previous;

    printf("%d", m_variable);
}
```
```
/Users/piers/a/test.cpp:31:18: warning: Access to field 'm_variable' results in
      a dereference of a null pointer
    printf("%d", m_variable);
                 ^
```
(the second warning about an unused variable can be safely ignored and only exists to keep the case simple)

**Comment by philnik777** [comment]
I can't reproduce this: https://godbolt.org/z/8MKbM5GPM
